web_poet/overrides.py

```diff
@@ -0,0 +1,146 @@
+import importlib
+import importlib.util
+import pkgutil
+import sys
+from dataclasses import dataclass, field
+from typing import Iterable, Union, List, Callable, Dict, Any
+
+from url_matcher import Patterns
+
+
+OVERRIDES_NAMESPACES_KEY = "_overrides_namespaces_"
+
+
+@dataclass(frozen=True)
+class HandleUrlsSpec:
+    """Meta information used by the :py:func:`web_poet.handle_urls` decorator"""
+    patterns: Patterns
+    overrides: Callable
+    meta: Dict[str, Any] = field(default_factory=dict)
+
+
+@dataclass(frozen=True)
+class OverrideRule:
+    """A single override rule. Specify when a page object should be used instead of another"""
+    for_patterns: Patterns
+    use: Callable
+    instead_of: Callable
+    meta: Dict[str, Any] = field(default_factory=dict)
+
+
+def _as_list(value: Union[str, Iterable[str], None]) -> List[str]:
+    """
+    >>> _as_list(None)
+    []
+    >>> _as_list("foo")
+    ['foo']
+    >>> _as_list(["foo", "bar"])
+    ['foo', 'bar']
+    """
+    if value is None:
+        return []
+    if isinstance(value, str):
+        return [value]
+    return list(value)
+
+
+def handle_urls(include: Union[str, Iterable[str]],
+                overrides: Callable,
+                *,
+                exclude: Union[str, Iterable[str], None] = None,
+                priority: int = 500,
+                namespace: str = "",
+                **kwargs
+                ):
```

Comment:
What do you think about using classes instead of string namespaces within a dictionary?
This is what I have in mind, it'd probably be easier to explain with some code:

```py
class PageObjectRegistry:
    def __init__(self):
        self.data = {}
    
    def handles_urls(self, include: ..., ...):
        # no need to do magic to get handle_urls_dict and namespace_dict,
        # use self.data instead of namespace_dict


default_registry = PageObjectRegistry()

# expose a shortcut for the default registry, so that the usage is the same:
handle_urls = default_registry.handle_urls
```

Reply:
Good point! I like the idea! 💡 as having them as classes would mean they can be extended to handle even more features later on.

I refactored the namespaces into a `PageObjectRegistry` instead in https://github.com/scrapinghub/web-poet/pull/16/commits/f5cffefd521e17c419ef2db004d0fbc717417aa8.

Reply:
I was thinking of renaming `PageObjectRegistry` to something else, as it might confuse the users later on when they use `scrapy-poet`. However, the terminology seems apt at the moment.

Reply:
Handled this on https://github.com/scrapinghub/web-poet/pull/16/commits/f5cffefd521e17c419ef2db004d0fbc717417aa8